torch/distributed/pipelining/PipelineStage.py

```diff
@@ -464,8 +453,17 @@ def forward_maybe_with_nosync(self, *args, **kwargs):
         return out_val
 
     def backward_maybe_with_nosync(self, bwd_kwargs: Dict, bwd_chunk_id: int):
+        """
+        Whether using PP with FSDP or DDP, there are some runtime differences between the last backward step and the
+        other steps.  Namely, we need to accumulate gradients on previous steps and reduce them on the last step, but
+        there are additional state-variables and performance considerations depending on the data parallelism used.
+        This helper should adapt any pipeline parallel schedule to work with common/supported data parallel libraries.
+        """
+        last_backward = bwd_chunk_id == self.chunks - 1
```

Comment:
isn't this broken if the schedule tries to do the bwd chunks in the reverse order as @H-Huang was proposing for bfs?

Reply:
Discussed and not an issue now.

Reply:
uh.. it is still an issue! in the previous code, the schedule was controlling 'data parallel mode' so it was the schedule's responsibility to make sure that it got called correctly.

unless we actively assert that schedules always call bwd_chunks in linear order (0..N), which we don't, then this adds a little risk exposure.  Should we at least assert this?

Reply:
You caught me. Maybe we still need a `seen_bwd_chunks` counter in `PipelineStage`. And:
```
last_backward = seen_bwd_chunks == self.chunks - 1
```

Reply:
> last_backward = seen_bwd_chunks == self.chunks - 1

If we go this route then we wouldn't need to pass bwd_chunk_id into `backward_maybe_with_nosync` right?